lib/resources.js

```diff
@@ -524,6 +549,101 @@ module.exports = function (app, _private = null) {
       : {}
   }
 
+  /**
+   * Given an array of status ids (e.g. "status:a", "status:na") returns the
+   * subset of statuses that are relevant in ReCAP
+   */
+  const recapStatuses = (statuses) => {
+    return statuses
+      .filter((status) => ['status:a', 'status:na'].includes(status))
+  }
+
+  /**
+   *  Builds a big complicated ES filter to allow us to filter items by status,
+   *  but override the indexed status for ReCAP items, whose status has been
+   *  retrieved from SCSB.
+   *
+   *  @param {string[]} statuses - An array of statuses to filter on
+   *  @param {string[]} unavailableRecapBarcodes - An array of item barcodes
+   *    known to be unavailble
+   *
+   *  Returns an ES filter that matches the desired statuses, but also uses
+   *  the known unavailable items to override indexed item statuses for ReCAP
+   *  items (because ReCAP is the authority for status of off-site items).
+   *  Essentially, the criteria is for matching an item is:
+   *
+   *   - if on-site (non-ReCAP):
+   *     - has a matching indexed status
+   *   - if off-site:
+   *     - if filtering on status:na
+   *       - item barcode must be in unavailableRecapBarcodes
+   *     - if filtering on status:a:
+   *       - item barcode must NOT be in unavailableRecapBarcodes
+   */
+  const itemStatusFilterWithUnavailableRecapItems = (statuses, unavailableRecapBarcodes) => {
+    // First, let's set up some common clauses:
+
+    // Item is in ReCAP:
+    const itemIsRecapClause = {
+      regexp: { 'items.holdingLocation.id': 'loc:rc.*' }
+    }
+    // Item's indexed status matches one of the filtered statuses:
+    const itemHasIndexedStatusClause = {
+      terms: { 'items.status.id': statuses }
+    }
+    // Item is marked Unavailable in SCSB:
+    const itemIsUnavailableInRecapClause = {
+      script: {
```

Comment:
does this need to be nested like this?

Reply:
Yes. I'm not sure why, but [that's how one adds script queries](https://www.elastic.co/guide/en/elasticsearch/reference/5.3/query-dsl-script-query.html)